[Bug] write consumer file exception will cause streaming read   task with consumerId error

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.


### Paimon version

0.5-SNAPSHOT

### Compute Engine

Flink 1.17

### Minimal reproduce step

**STEP 1:when checkpoint complete, org.apache.paimon.consumer.ConsumerManager.recordConsumer report error:**

java.lang.IllegalArgumentException: Self-suppression not permitted
	at java.lang.Throwable.addSuppressed(Throwable.java:1043) ~[?:1.8.0_191]
	at org.apache.paimon.consumer.ConsumerManager.recordConsumer(ConsumerManager.java:61) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at org.apache.paimon.table.source.InnerStreamTableScanImpl.notifyCheckpointComplete(InnerStreamTableScanImpl.java:240) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at java.util.OptionalLong.ifPresent(OptionalLong.java:142) ~[?:1.8.0_191]
	at org.apache.paimon.flink.source.operator.MonitorFunction.notifyCheckpointComplete(MonitorFunction.java:207) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
Caused by: java.io.IOException: Could not get block locations. Source file "/xxxx/jarvis_run_history_compute/consumer/consumer-sr_jarvis_run_history_compute" - Aborting...block==null
	at org.apache.hadoop.hdfs.DataStreamer.setupPipelineForAppendOrRecovery(DataStreamer.java:1508) ~[hadoop-hdfs-client-3.3.1.jar:?]


**STEP2:The conent of consumer file is blank**

**STEP3:streaming read   task with consumerId error:**

2023-10-17 10:39:18,528 WARN  org.apache.flink.runtime.taskmanager.Task                    [] - Source: default_catalog.default_database.jarvis_run_history_compute-Monitor (1/1)#3 (c4e704687898bfd0f5d0bd2644629fad_bc764cd8ddf7a0cff126f51c16239658_0_3) switched from RUNNING to FAILED with failure cause:
java.io.UncheckedIOException: org.apache.paimon.shade.jackson2.com.fasterxml.jackson.databind.exc.MismatchedInputException: No content to map due to end-of-input
 at [Source: (String)""; line: 1, column: 0]
	at org.apache.paimon.utils.JsonSerdeUtil.fromJson(JsonSerdeUtil.java:59) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at org.apache.paimon.consumer.Consumer.fromJson(Consumer.java:54) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at org.apache.paimon.consumer.Consumer.fromPath(Consumer.java:64) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at org.apache.paimon.consumer.ConsumerManager.consumer(ConsumerManager.java:53) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at org.apache.paimon.table.source.AbstractInnerTableScan.createStartingScanner(AbstractInnerTableScan.java:88) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at org.apache.paimon.table.source.InnerStreamTableScanImpl.plan(InnerStreamTableScanImpl.java:81) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at org.apache.paimon.flink.source.operator.MonitorFunction.run(MonitorFunction.java:180) ~[paimon-flink-1.17-0.5-SNAPSHOT.jar:0.5-SNAPSHOT]
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:110) ~[flink-dist-1.17.0.jar:1.17.0]
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:67) ~[flink-dist-1.17.0.jar:1.17.0]


### What doesn't meet your expectations?


streaming read  task failed and kept restarting . And it  lost consumer offset info .

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java
+++ b/paimon-common/src/main/java/org/apache/paimon/fs/FileIO.java
@@ -42,9 +42,9 @@ import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.ServiceLoader;
 import java.util.Set;
-import java.util.UUID;
 import java.util.stream.Collectors;
 
 import static org.apache.paimon.fs.FileIOUtils.checkAccess;
@@ -204,7 +204,7 @@ public interface FileIO extends Serializable {
             return false;
         }
 
-        Path tmp = new Path(path.getParent(), "." + path.getName() + UUID.randomUUID());
+        Path tmp = path.createTempPath();
         boolean success = false;
         try {
             try (PositionOutputStream out = newOutputStream(tmp, false)) {
@@ -221,6 +221,43 @@ public interface FileIO extends Serializable {
         }
 
         return success;
+    }
+
+    /**
+     * Overwrite file by content atomically, different {@link FileIO}s have different atomic
+     * implementations.
+     */
+    default void overwriteFileUtf8(Path path, String content) throws IOException {
+        try (PositionOutputStream out = newOutputStream(path, true)) {
+            OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+            writer.write(content);
+            writer.flush();
+        }
+    }
+
+    /** Read file from {@link #overwriteFileUtf8} file. */
+    default Optional<String> readOverwrittenFileUtf8(Path path) throws IOException {
+        int retryNumber = 0;
+        IOException exception = null;
+        while (retryNumber++ < 5) {
+            try {
+                if (!exists(path)) {
+                    return Optional.empty();
+                }
+
+                return Optional.of(readFileUtf8(path));
+            } catch (IOException e) {
+                if (e.getClass()
+                        .getName()
+                        .endsWith("org.apache.hadoop.fs.s3a.RemoteFileChangedException")) {
+                    exception = e;
+                } else {
+                    throw e;
+                }
+            }
+        }
+
+        throw exception;
     }
 
     // -------------------------------------------------------------------------

--- a/paimon-common/src/main/java/org/apache/paimon/fs/Path.java
+++ b/paimon-common/src/main/java/org/apache/paimon/fs/Path.java
@@ -28,6 +28,7 @@ import org.apache.paimon.utils.StringUtils;
 import java.io.Serializable;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.util.UUID;
 import java.util.regex.Pattern;
 
 /**
@@ -316,6 +317,11 @@ public class Path implements Comparable<Path>, Serializable {
         String path = uri.getPath();
         int slash = path.lastIndexOf(SEPARATOR);
         return path.substring(slash + 1);
+    }
+
+    /** Create a temporary path (to be used as a copy) for this path. */
+    public Path createTempPath() {
+        return new Path(getParent(), String.format(".%s.%s.tmp", getName(), UUID.randomUUID()));
     }
 
     /**

--- a/paimon-common/src/main/java/org/apache/paimon/fs/hadoop/HadoopFileIO.java
+++ b/paimon-common/src/main/java/org/apache/paimon/fs/hadoop/HadoopFileIO.java
@@ -35,7 +35,6 @@ import org.apache.hadoop.fs.Options;
 
 import java.io.IOException;
 import java.io.OutputStreamWriter;
-import java.io.UncheckedIOException;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.nio.charset.StandardCharsets;
@@ -122,6 +121,14 @@ public class HadoopFileIO implements FileIO {
         org.apache.hadoop.fs.Path hadoopSrc = path(src);
         org.apache.hadoop.fs.Path hadoopDst = path(dst);
         return getFileSystem(hadoopSrc).rename(hadoopSrc, hadoopDst);
+    }
+
+    @Override
+    public void overwriteFileUtf8(Path path, String content) throws IOException {
+        boolean success = tryAtomicOverwriteViaRename(path, content);
+        if (!success) {
+            FileIO.super.overwriteFileUtf8(path, content);
+        }
     }
 
     private org.apache.hadoop.fs.Path path(Path path) {
@@ -340,8 +347,6 @@ public class HadoopFileIO implements FileIO {
                 OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
                 writer.write(content);
                 writer.flush();
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
             }
 
             renameMethod.invoke(

--- a/paimon-common/src/main/java/org/apache/paimon/fs/hadoop/HadoopFileIO.java
+++ b/paimon-common/src/main/java/org/apache/paimon/fs/hadoop/HadoopFileIO.java
@@ -26,12 +26,20 @@ import org.apache.paimon.fs.Path;
 import org.apache.paimon.fs.PositionOutputStream;
 import org.apache.paimon.fs.SeekableInputStream;
 import org.apache.paimon.hadoop.SerializableConfiguration;
+import org.apache.paimon.utils.ReflectionUtils;
 
 import org.apache.hadoop.fs.FSDataInputStream;
 import org.apache.hadoop.fs.FSDataOutputStream;
 import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Options;
 
 import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.UncheckedIOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.atomic.AtomicReference;
 
 /** Hadoop {@link FileIO}. */
 public class HadoopFileIO implements FileIO {
@@ -288,6 +296,89 @@ public class HadoopFileIO implements FileIO {
         @Override
         public long getModificationTime() {
             return status.getModificationTime();
+        }
+    }
+
+    // ============================== extra methods ===================================
+
+    private transient volatile AtomicReference<Method> renameMethodRef;
+
+    public boolean tryAtomicOverwriteViaRename(Path dst, String content) throws IOException {
+        org.apache.hadoop.fs.Path hadoopDst = path(dst);
+        FileSystem fs = getFileSystem(hadoopDst);
+
+        if (renameMethodRef == null) {
+            synchronized (this) {
+                if (renameMethodRef == null) {
+                    Method method;
+                    // Implementation in FileSystem is incorrect, not atomic, Object Storage like S3
+                    // and OSS not override it
+                    // DistributedFileSystem and ViewFileSystem override the rename method to public
+                    // and implement correct renaming
+                    try {
+                        method = ReflectionUtils.getMethod(fs.getClass(), "rename", 3);
+                    } catch (NoSuchMethodException e) {
+                        method = null;
+                    }
+                    renameMethodRef = new AtomicReference<>(method);
+                }
+            }
+        }
+
+        Method renameMethod = renameMethodRef.get();
+        if (renameMethod == null) {
+            return false;
+        }
+
+        boolean renameDone = false;
+
+        // write tempPath
+        Path tempPath = dst.createTempPath();
+        org.apache.hadoop.fs.Path hadoopTemp = path(tempPath);
+        try {
+            try (PositionOutputStream out = newOutputStream(tempPath, false)) {
+                OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+                writer.write(content);
+                writer.flush();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+
+            renameMethod.invoke(
+                    fs, hadoopTemp, hadoopDst, new Options.Rename[] {Options.Rename.OVERWRITE});
+            renameDone = true;
+            // TODO: this is a workaround of HADOOP-16255 - remove this when HADOOP-16255 is
+            // resolved
+            tryRemoveCrcFile(hadoopTemp);
+            return true;
+        } catch (InvocationTargetException | IllegalAccessException e) {
+            throw new IOException(e);
+        } finally {
+            if (!renameDone) {
+                deleteQuietly(tempPath);
+            }
+        }
+    }
+
+    /** @throws IOException if a fatal exception occurs. Will try to ignore most exceptions. */
+    @SuppressWarnings("CatchMayIgnoreException")
+    private void tryRemoveCrcFile(org.apache.hadoop.fs.Path path) throws IOException {
+        try {
+            final org.apache.hadoop.fs.Path checksumFile =
+                    new org.apache.hadoop.fs.Path(
+                            path.getParent(), String.format(".%s.crc", path.getName()));
+
+            if (fs.exists(checksumFile)) {
+                // checksum file exists, deleting it
+                fs.delete(checksumFile, true); // recursive=true
+            }
+        } catch (Throwable t) {
+            if (t instanceof VirtualMachineError
+                    || t instanceof ThreadDeath
+                    || t instanceof LinkageError) {
+                throw t;
+            }
+            // else, ignore - we are removing crc file as "best-effort"
         }
     }
 }

--- a/paimon-common/src/main/java/org/apache/paimon/utils/ReflectionUtils.java
+++ b/paimon-common/src/main/java/org/apache/paimon/utils/ReflectionUtils.java
@@ -47,4 +47,21 @@ public class ReflectionUtils {
             throws InvocationTargetException, IllegalAccessException {
         return (T) method.invoke(null, args);
     }
+
+    public static Method getMethod(Class<?> clz, String methodName, int argSize)
+            throws NoSuchMethodException {
+        Method method = null;
+        Method[] methods = clz.getMethods();
+        for (Method m : methods) {
+            if (methodName.equals(m.getName()) && m.getParameterTypes().length == argSize) {
+                method = m;
+                break;
+            }
+        }
+
+        if (method == null) {
+            throw new NoSuchMethodException(methodName);
+        }
+        return method;
+    }
 }

--- a/paimon-common/src/test/java/org/apache/paimon/fs/FileIOBehaviorTestBase.java
+++ b/paimon-common/src/test/java/org/apache/paimon/fs/FileIOBehaviorTestBase.java
@@ -227,7 +227,7 @@ public abstract class FileIOBehaviorTestBase {
     //  Utilities
     // ------------------------------------------------------------------------
 
-    private static String randomName() {
+    protected static String randomName() {
         return StringUtils.getRandomString(RND, 16, 16, 'a', 'z');
     }
 

--- a/paimon-common/src/test/java/org/apache/paimon/fs/FileIOTest.java
+++ b/paimon-common/src/test/java/org/apache/paimon/fs/FileIOTest.java
@@ -26,6 +26,8 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
 import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -55,5 +57,53 @@ public class FileIOTest {
                         new Path("require-options://" + tempDir.toString()),
                         CatalogContext.create(options));
         assertThat(fileIO).isInstanceOf(RequireOptionsFileIOLoader.MyFileIO.class);
+    }
+
+    public static void testOverwriteFileUtf8(Path file, FileIO fileIO) throws InterruptedException {
+        AtomicReference<Exception> exception = new AtomicReference<>();
+        final int max = 10;
+
+        Thread writeThread =
+                new Thread(
+                        () -> {
+                            for (int i = 0; i <= max; i++) {
+                                try {
+                                    fileIO.overwriteFileUtf8(file, "" + i);
+                                    Thread.sleep(100);
+                                } catch (Exception e) {
+                                    exception.set(e);
+                                    return;
+                                }
+                            }
+                        });
+
+        Thread readThread =
+                new Thread(
+                        () -> {
+                            while (true) {
+                                try {
+                                    Optional<String> ret = fileIO.readOverwrittenFileUtf8(file);
+                                    if (!ret.isPresent()) {
+                                        continue;
+                                    }
+
+                                    int value = Integer.parseInt(ret.get());
+                                    if (value == max) {
+                                        return;
+                                    }
+                                } catch (Exception e) {
+                                    exception.set(e);
+                                    return;
+                                }
+                            }
+                        });
+
+        writeThread.start();
+        readThread.start();
+
+        writeThread.join();
+        readThread.join();
+
+        assertThat(exception.get()).isNull();
     }
 }

--- a/paimon-common/src/test/java/org/apache/paimon/fs/HdfsBehaviorTest.java
+++ b/paimon-common/src/test/java/org/apache/paimon/fs/HdfsBehaviorTest.java
@@ -25,10 +25,14 @@ import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hdfs.MiniDFSCluster;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
 import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicReference;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assumptions.assumeThat;
 
 /** Behavior tests for HDFS. */
@@ -83,5 +87,61 @@ class HdfsBehaviorTest extends FileIOBehaviorTestBase {
     @Override
     protected Path getBasePath() {
         return basePath;
+    }
+
+    @Test
+    public void testAtomicWrite() throws IOException {
+        Path file = new Path(getBasePath(), randomName());
+        fs.tryAtomicOverwriteViaRename(file, "Hi");
+        assertThat(fs.readFileUtf8(file)).isEqualTo("Hi");
+
+        fs.tryAtomicOverwriteViaRename(file, "Hello");
+        assertThat(fs.readFileUtf8(file)).isEqualTo("Hello");
+    }
+
+    @Test
+    public void testAtomicWriteMultipleThreads() throws InterruptedException, IOException {
+        Path file = new Path(getBasePath(), randomName());
+        AtomicReference<Exception> exception = new AtomicReference<>();
+        final int max = 10;
+        fs.tryAtomicOverwriteViaRename(file, "0");
+
+        Thread writeThread =
+                new Thread(
+                        () -> {
+                            for (int i = 1; i <= max; i++) {
+                                try {
+                                    fs.tryAtomicOverwriteViaRename(file, "" + i);
+                                    Thread.sleep(100);
+                                } catch (Exception e) {
+                                    exception.set(e);
+                                    return;
+                                }
+                            }
+                        });
+
+        Thread readThread =
+                new Thread(
+                        () -> {
+                            while (true) {
+                                try {
+                                    int value = Integer.parseInt(fs.readFileUtf8(file));
+                                    if (value == max) {
+                                        return;
+                                    }
+                                } catch (Exception e) {
+                                    exception.set(e);
+                                    return;
+                                }
+                            }
+                        });
+
+        writeThread.start();
+        readThread.start();
+
+        writeThread.join();
+        readThread.join();
+
+        assertThat(exception.get()).isNull();
     }
 }

--- a/paimon-common/src/test/java/org/apache/paimon/fs/HdfsBehaviorTest.java
+++ b/paimon-common/src/test/java/org/apache/paimon/fs/HdfsBehaviorTest.java
@@ -30,7 +30,6 @@ import org.junit.jupiter.api.io.TempDir;
 
 import java.io.File;
 import java.io.IOException;
-import java.util.concurrent.atomic.AtomicReference;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assumptions.assumeThat;
@@ -100,48 +99,7 @@ class HdfsBehaviorTest extends FileIOBehaviorTestBase {
     }
 
     @Test
-    public void testAtomicWriteMultipleThreads() throws InterruptedException, IOException {
-        Path file = new Path(getBasePath(), randomName());
-        AtomicReference<Exception> exception = new AtomicReference<>();
-        final int max = 10;
-        fs.tryAtomicOverwriteViaRename(file, "0");
-
-        Thread writeThread =
-                new Thread(
-                        () -> {
-                            for (int i = 1; i <= max; i++) {
-                                try {
-                                    fs.tryAtomicOverwriteViaRename(file, "" + i);
-                                    Thread.sleep(100);
-                                } catch (Exception e) {
-                                    exception.set(e);
-                                    return;
-                                }
-                            }
-                        });
-
-        Thread readThread =
-                new Thread(
-                        () -> {
-                            while (true) {
-                                try {
-                                    int value = Integer.parseInt(fs.readFileUtf8(file));
-                                    if (value == max) {
-                                        return;
-                                    }
-                                } catch (Exception e) {
-                                    exception.set(e);
-                                    return;
-                                }
-                            }
-                        });
-
-        writeThread.start();
-        readThread.start();
-
-        writeThread.join();
-        readThread.join();
-
-        assertThat(exception.get()).isNull();
+    public void testAtomicWriteMultipleThreads() throws InterruptedException {
+        FileIOTest.testOverwriteFileUtf8(new Path(getBasePath(), randomName()), fs);
     }
 }

--- a/paimon-core/src/main/java/org/apache/paimon/consumer/Consumer.java
+++ b/paimon-core/src/main/java/org/apache/paimon/consumer/Consumer.java
@@ -26,8 +26,9 @@ import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonCre
 import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonGetter;
 import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
 
+import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.util.Optional;
-import java.util.concurrent.TimeUnit;
 
 /** Consumer which contains next snapshot. */
 public class Consumer {
@@ -58,31 +59,10 @@ public class Consumer {
     }
 
     public static Optional<Consumer> fromPath(FileIO fileIO, Path path) {
-        // Consumer updating uses FileIO.newOutputStream(..., overwrite).
-        // But this API may have some intermediate state, the file maybe empty
-        // So retry here to avoid exception when the file is intermediate state
-        int retryNumber = 0;
-        Exception exception = null;
-        while (retryNumber++ < READ_CONSUMER_RETRY_NUM) {
-            try {
-                if (!fileIO.exists(path)) {
-                    return Optional.empty();
-                }
-
-                String json = fileIO.readFileUtf8(path);
-                return Optional.of(Consumer.fromJson(json));
-            } catch (Exception e) {
-                exception = e;
-            }
-
-            try {
-                TimeUnit.MILLISECONDS.sleep(READ_CONSUMER_RETRY_INTERVAL);
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new RuntimeException(e);
-            }
+        try {
+            return fileIO.readOverwrittenFileUtf8(path).map(Consumer::fromJson);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
         }
-
-        throw new RuntimeException("Fails to read snapshot from path " + path, exception);
     }
 }

--- a/paimon-core/src/main/java/org/apache/paimon/consumer/ConsumerManager.java
+++ b/paimon-core/src/main/java/org/apache/paimon/consumer/ConsumerManager.java
@@ -20,15 +20,11 @@ package org.apache.paimon.consumer;
 
 import org.apache.paimon.fs.FileIO;
 import org.apache.paimon.fs.Path;
-import org.apache.paimon.fs.PositionOutputStream;
-import org.apache.paimon.fs.hadoop.HadoopFileIO;
 import org.apache.paimon.utils.DateTimeUtils;
 
 import java.io.IOException;
-import java.io.OutputStreamWriter;
 import java.io.Serializable;
 import java.io.UncheckedIOException;
-import java.nio.charset.StandardCharsets;
 import java.time.LocalDateTime;
 import java.util.HashMap;
 import java.util.List;
@@ -60,26 +56,8 @@ public class ConsumerManager implements Serializable {
     }
 
     public void resetConsumer(String consumerId, Consumer consumer) {
-        Path path = consumerPath(consumerId);
-        if (fileIO instanceof HadoopFileIO) {
-            // For HDFS, try to use rename with overwrite
-            try {
-                boolean success =
-                        ((HadoopFileIO) fileIO)
-                                .tryAtomicOverwriteViaRename(path, consumer.toJson());
-                if (success) {
-                    return;
-                }
-            } catch (IOException e) {
-                throw new RuntimeException(e);
-            }
-        }
-
-        // For Non-HDFS, just use write overwrite
-        try (PositionOutputStream out = fileIO.newOutputStream(consumerPath(consumerId), true)) {
-            OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
-            writer.write(consumer.toJson());
-            writer.flush();
+        try {
+            fileIO.overwriteFileUtf8(consumerPath(consumerId), consumer.toJson());
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }

--- a/paimon-core/src/main/java/org/apache/paimon/consumer/ConsumerManager.java
+++ b/paimon-core/src/main/java/org/apache/paimon/consumer/ConsumerManager.java
@@ -21,6 +21,7 @@ package org.apache.paimon.consumer;
 import org.apache.paimon.fs.FileIO;
 import org.apache.paimon.fs.Path;
 import org.apache.paimon.fs.PositionOutputStream;
+import org.apache.paimon.fs.hadoop.HadoopFileIO;
 import org.apache.paimon.utils.DateTimeUtils;
 
 import java.io.IOException;
@@ -59,6 +60,22 @@ public class ConsumerManager implements Serializable {
     }
 
     public void resetConsumer(String consumerId, Consumer consumer) {
+        Path path = consumerPath(consumerId);
+        if (fileIO instanceof HadoopFileIO) {
+            // For HDFS, try to use rename with overwrite
+            try {
+                boolean success =
+                        ((HadoopFileIO) fileIO)
+                                .tryAtomicOverwriteViaRename(path, consumer.toJson());
+                if (success) {
+                    return;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        // For Non-HDFS, just use write overwrite
         try (PositionOutputStream out = fileIO.newOutputStream(consumerPath(consumerId), true)) {
             OutputStreamWriter writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
             writer.write(consumer.toJson());

--- a/paimon-spark/paimon-spark-common/pom.xml
+++ b/paimon-spark/paimon-spark-common/pom.xml
@@ -180,6 +180,13 @@ under the License.
         </dependency>
         <dependency>
             <groupId>org.apache.paimon</groupId>
+            <artifactId>paimon-common</artifactId>
+            <version>${project.version}</version>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.paimon</groupId>
             <artifactId>paimon-hive-common</artifactId>
             <version>${project.version}</version>
             <scope>test</scope>

--- a/paimon-spark/paimon-spark-common/src/test/java/org/apache/paimon/spark/SparkS3ITCase.java
+++ b/paimon-spark/paimon-spark-common/src/test/java/org/apache/paimon/spark/SparkS3ITCase.java
@@ -18,7 +18,11 @@
 
 package org.apache.paimon.spark;
 
+import org.apache.paimon.catalog.CatalogContext;
+import org.apache.paimon.fs.FileIO;
+import org.apache.paimon.fs.FileIOTest;
 import org.apache.paimon.fs.Path;
+import org.apache.paimon.options.Options;
 import org.apache.paimon.s3.MinioTestContainer;
 import org.apache.paimon.testutils.junit.parameterized.ParameterizedTestExtension;
 import org.apache.paimon.testutils.junit.parameterized.Parameters;
@@ -32,6 +36,7 @@ import org.junit.jupiter.api.TestTemplate;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
+import java.io.IOException;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
@@ -46,12 +51,14 @@ public class SparkS3ITCase {
     @RegisterExtension
     public static final MinioTestContainer MINIO_CONTAINER = new MinioTestContainer();
 
+    private static Path warehousePath;
+
     private static SparkSession spark = null;
 
     @BeforeAll
     public static void startMetastoreAndSpark() {
         String path = MINIO_CONTAINER.getS3UriForDefaultBucket() + "/" + UUID.randomUUID();
-        Path warehousePath = new Path(path);
+        warehousePath = new Path(path);
         spark = SparkSession.builder().master("local[2]").getOrCreate();
         spark.conf().set("spark.sql.catalog.paimon", SparkCatalog.class.getName());
         spark.conf().set("spark.sql.catalog.paimon.warehouse", warehousePath.toString());
@@ -83,7 +90,7 @@ public class SparkS3ITCase {
 
     @AfterEach
     public void afterEach() {
-        spark.sql("DROP TABLE T");
+        spark.sql("DROP TABLE IF EXISTS T");
     }
 
     @TestTemplate
@@ -96,5 +103,14 @@ public class SparkS3ITCase {
         spark.sql("INSERT INTO T VALUES (1, 2, '3')").collectAsList();
         List<Row> rows = spark.sql("SELECT * FROM T").collectAsList();
         assertThat(rows.toString()).isEqualTo("[[1,2,3]]");
+    }
+
+    @TestTemplate
+    public void testS3AtomicWriteMultipleThreads() throws InterruptedException, IOException {
+        Path file = new Path(warehousePath, UUID.randomUUID().toString());
+        Options options = new Options();
+        MINIO_CONTAINER.getS3ConfigOptions().forEach(options::setString);
+        FileIO fileIO = FileIO.get(file, CatalogContext.create(options));
+        FileIOTest.testOverwriteFileUtf8(file, fileIO);
     }
 }

